### PR TITLE
Implement FinalizeBlock changes for validator set update protocol txs

### PR DESCRIPTION
**Pull request by sug0**
Based on #676

Implements item no. 7 of https://github.com/anoma/namada/issues/242

This PR will track all changes related with validator set update tallies.

- [x] https://github.com/anoma/namada/pull/684
- [x] https://github.com/anoma/namada/pull/690
- [x] https://github.com/anoma/namada/pull/731
- [x] https://github.com/anoma/namada/pull/739

For convenience, #753 has been merged in, so that PR should be reviewed first, and diffed against.

### CODE REVIEW

File: shared/src/ledger/protocol/transactions/ethereum_events/mod.rs [hunk 1]
```diff
@@ -442,4 +418,61 @@ mod tests {
              voting power so far"
         );
     }
+
+    #[test]
+    /// Assert we don't return anything if we try to get the votes for an empty
+    /// vec of events
+    pub fn test_get_votes_for_events_empty() {
```
Comment (batconjurer) [inline shared/src/ledger/protocol/transactions/ethereum_events/mod.rs:425]: We should have a test where a message has multiple copies of a sig from the same validator.

**Review by batconjurer — CHANGES_REQUESTED** [review]

**Review by james-chf — APPROVED** [review]

File: shared/src/ledger/protocol/transactions/ethereum_events/mod.rs (hunk 1 above)
Comment (sug0) [reply shared/src/ledger/protocol/transactions/ethereum_events/mod.rs:425]: that should be written for #760

File: shared/src/types/storage.rs [hunk 2]
```diff
@@ -688,6 +688,32 @@ impl KeySeg for BlockHeight {
...
+        string
+            .split_once('=')
+            .and_then(|(prefix, epoch)| (prefix == "E").then(|| epoch))
+            .ok_or_else(|| Error::Temporary {
+                error: format!("Invalid epoch prefix on key: {string}"),
+            })
+            .and_then(|epoch| {
+                epoch.parse::<u64>().map_err(|e| Error::Temporary {
+                    error: format!("Unexpected epoch value {epoch}, {e}"),
+                })
```
[the remark this answers is not in the document]
Comment (sug0) [reply shared/src/types/storage.rs:702]: seems more like a catch-all error type, when you're feeling lazy to write a new variant

**Review by batconjurer — APPROVED** [review]

File: shared/src/types/storage.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (james-chf) [reply shared/src/types/storage.rs]: we should (maybe) have an `Error::Other` variant or smth (or just use `eyre` more)

**Pull request merged**